keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+
+As a machine-learning researcher from the [Story 2](#story-2), I want to mark the job
+as successful if 50% of all indexes succeed because I can not know how many indexes 
+will have in advance.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      count: 50%
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+#### Story 4
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and continue running the rest of the pods if some indexes succeed
+because I set the minimum required value for sampling in the `.successPolicy.onSucceededPods.count` and
+perform the same simulation in all indexes.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      count: 5
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.onSucceededPods.indexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `indexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
```

Comment:
Are Indexes and Count mutually exclusive?  I think so based on use-cases and how they would interact, but I'd like to be sure.

If mutually exclusive, is there a need to have different potential conditions as successPolicy (array) or is a single one sufficient?

Reply:
And if mutually exclusive, make a discriminated union and specify which one is intended.

Reply:
> Are Indexes and Count mutually exclusive? I think so based on use-cases and how they would interact, but I'd like to be sure.

Yes, the `indexes` and `count` are mutually exclusive.

> If mutually exclusive, is there a need to have different potential conditions as successPolicy (array) or is a single one sufficient?

Does "different potential conditions as successPolicy (array)" mean the following situation?

```yaml
...
  successPolicy:
    - onSucceededPods:
        count: 5
        whenTargetAchieved: terminate
    - onSucceededPods:
        indexes: "0,1"
        whenTargetAchieved: continue
...
```

If so, I think a single one is sufficient for now since there is not any effective use case for making the successPolicy conditions array in the `onSucceededPods` condition.
However, as I described in https://github.com/kubernetes/enhancements/pull/4062#discussion_r1460403114, since it is possible to add new conditions in the future, making the succesPolicy condition array like the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/) would be nice to have.

Reply:
@deads2k Oh, I find an effective use case that the `indexes` and `count` are not mutually exclusive like this.
This is a similar user story to the user story 4.

```yaml
spec:
  completions: 10
  successPolicy:
    - onSucceededPods:
        indexes: "1-9"
        count: 5
        whenTargetAchieved: continue
    - onSucceededPods:
        indexes: "0"
        count: 1
        whenTargetAchieved: continue
```

In the above situations, the index `0` takes on the role of a leader, and the `1-9` indexes take on the role of `worker`. At this time, we want to mark the job as successful (completed) when the leader succeeds, and workers with the minimum required value of `5` for sampling succeed.

@mimowo I explained that the `indexes` and `count` are mutually exclusive in https://github.com/kubernetes/enhancements/pull/4062#discussion_r1453287814. However, I find an effective use case for allowing both (`indexes` and `count`) like the above. WDYT?

Reply:
> @mimowo I explained that the `indexes` and `count` are mutually exclusive in [#4062 (comment)](https://github.com/kubernetes/enhancements/pull/4062#discussion_r1453287814). However, I find an effective use case for allowing both (`indexes` and `count`) like the above. WDYT?

Sure, if there is a legitimate use case we should support it. I would also suggest to mention the use case in the User stories section. 

One consideration from me; do we need different actions `whenTargetAchieved`, for different targets represented by `onSucceededPods`? What if both targets match, which action is chosen there. Food for thought, but maybe this API is enough:
```yaml

spec:
  completions: 10
  successPolicy:
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
    - target:
        succeededIndexes: "0"
        succeededcount: 1
      whenTargetAchieved: continue
```
I don't have a strong view, just a proposal. I just don't feel like we would ever have different actions (`whenTargetAchieved) depending on the success target reached.

Reply:
>  I would also suggest to mention the use case in the User stories section.

It makes sense.

> do we need different actions whenTargetAchieved, for different targets represented by onSucceededPods? 

I still think that we need different actions since we may want to add new policies like the following in the future:

```yaml
spec:
  completions: 10
  successPolicy:
    - whenTargetAchieved: terminate
       target:
         onSucceededPods:
           indexes: "0"
           count: 1
    - whenTargetAchieved: continue
       target:
         onPolicyA:
           paramA0: ""
```

@mimowo Do you think that we should not have extensibility for the future? Let me know what you think.

>  What if both targets match, which action is chosen there. 

Similar to the pod failure policy, I think that we can evaluate policies in order.

> the Pod failure policy rules you specify under spec.podFailurePolicy.rules are evaluated in order. Once a rule matches a Pod failure, the remaining rules are ignored. When no rule matches the Pod failure, the default handling applies.

https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy

Reply:
> @mimowo Do you think that we should not have extensibility for the future? Let me know what you think.

I think having different actions depending on the "success" target seems excessive, and we don't have a use case for this at the moment. 

I think we still could provide extensibility by introducing `whenTargetAchieved` at the lower level as override of the default at the higher level. 

I see partial analogy to pod failure policy, but there it was very clear from the beginning we would have different actions depending on the matched rule. 

Here, it looks more like a room for error. Say we have a default `whenTargetAchieved: terminate`, and the user has two targets, but only sets the `whenTargetAchieved` value for one, as `continue`. Then, it would be defaulted in the other target. As a result, it would have two different values. IMO this would rarely be the intention.

Reply:
> I think having different actions depending on the "success" target seems excessive, and we don't have a use case for this at the moment.

I agree.

> I think we still could provide extensibility by introducing whenTargetAchieved at the lower level as override of the default at the higher level (in the future).

It sounds reasonable to me. @mimowo Does this mean the following?

```yaml
spec:
  completions: 10
  successPolicy:
    whenTargetAchieved: terminate
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
        whenTargetAchieved: continue # <- we way add in the future if needed.
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
```

> I see partial analogy to pod failure policy, but there it was very clear from the beginning we would have different actions depending on the matched rule.

I see. This is the good study for defining API. Thanks!

Reply:
> Here, it looks more like a room for error. Say we have a default whenTargetAchieved: terminate, and the user has two targets, but only sets the whenTargetAchieved value for one, as continue. Then, it would be defaulted in the other target. As a result, it would have two different values. IMO this would rarely be the intention.

Indeed, there is a concern in my suggesting API like what you say.
So, I feel attracted to @mimowo 's suggested API.

@deads2k Let us know about your concerns as an API reviewer regarding the following API design.

We discussed in 

API Design: A

```yaml
spec:
  completions: 10
  successPolicy:
    - whenTargetAchieved: terminate
      target:
       onSucceededPods:
         indexes: "0"
         count: 1
    - whenTargetAchieved: continue
      target:
        onPolicyA:                      # <- we may add in the future if needed.
          paramA0: ""
```

API Design B:

```yaml
spec:
  completions: 10
  successPolicy:
    whenTargetAchieved: terminate
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
        whenTargetAchieved: continue # <- we may add in the future if needed.
    - target:
        succeededIndexes: "1-9"
        succeededCount: 5
        paramFoo: ""  # <- we may add in the future if needed.
```

Reply:
If the `OnSucceededPods` slice under `SuccessPolicy` is ANDed, then the `whenTargetAchieved` logically belongs peer to the slice under `SuccessPolicy`.  If the `OnSucceededPods` slice is ORed, then `whenTargetAchieved` logically belongs in the `OnSucceededPods` item.

While deep nesting is user-hostile in cases where it is not needed, this is why `clusterRoleSelectors` in RBAC ORs the label selectors, but label selectors themselves are ANDed.

Reply:
Thank you for the great suggestion!

> If the OnSucceededPods slice under SuccessPolicy is ANDed, then the whenTargetAchieved logically belongs peer to the slice under SuccessPolicy. If the OnSucceededPods slice is ORed, then whenTargetAchieved logically belongs in the OnSucceededPods item.
> 
> While deep nesting is user-hostile in cases where it is not needed, this is why clusterRoleSelectors in RBAC ORs the label selectors, but label selectors themselves are ANDed.

I meant ORed. So, we should adopt the `API design B` and move `whenTargetAchieved` under the target.

```yaml
spec:
  completions: 10
  successPolicy:
    - target: # <- this level is ORed
        succeededIndexes: "1-9" # <- this level is ANDed
        succeededCount: 5
        whenTargetAchieved: continue
    - target:
        succeededIndexes: "0"
        succeededCount: 1
        whenTargetAchieved: terminate
```

In the above yaml, an index `0` is a leader, and other indexes (`1-9`) are workers.

When the job meets the first target (with whenTargetAchieved=continue), we want to mark the job as succeeded and to continue running the job's pods for sampling.

When the job meets the second target (with whenTargetAchieved=terminate), we want to mark the job as succeeded and terminate the running the job's pods since the completed leader means that the job is finished.

I will update the API design with this story in this KEP.

Although as @mimowo says, this API has the below concern, let me document it in this KEP.

> Here, it looks more like a room for error. Say we have a default whenTargetAchieved: terminate, and the user has two targets, but only sets the whenTargetAchieved value for one, as continue. Then, it would be defaulted in the other target. As a result, it would have two different values. IMO this would rarely be the intention.

Reply:
@deads2k @mimowo Done.